### DISCUSSION TITLE
Having a visual oddball task during a survey-text trial

### DISCUSSION THREAD

**Discussion by avashipman** [jspsych/jsPsych#869] [Q&A] [answered]
Hi,

I want to design an experiment where the user has to do two tasks at once. The user will have to write a sentence in the given textbox while also completing a visual oddball task as a distractor. I tried making my own plugin by using the image-button-response and the survey-text plugins. However, I am not sure if this is the right way to approach it.

Is there any way that I can have the visual oddball task running while the survey-text is also running in one trial?

I appreciate any help that I can get!

**Comment by vijaymarupudi** [comment]
You'll need to combine the plugin code for this one. This might be relatively complex depending on how familiar you are with JavaScript. 

The higher level view is that you need to set up two `<div>` tags, both positioned how you'd like the two concurrent tasks to be positioned. Give them a unique id and acquire a reference in JavaScript. Then use that as the `display_element` instead of the default in the plugin code.

Finally, for the distractor response you don't care about (or the response you don't want to end the trial), find the appropriate `jsPsych.finishTrial()` call and comment it out. What you do with the data that plugin saves is up to you.

**Reply by avashipman** [reply]
Thank you for your quick response! This is so helpful!

I am very new to JavaScript, so this is a bit complicated for me. When I make the

tags, is that in the HTML code or the plugin code? When you say to use the unique id as the display_element, do you mean to replace display_element with the unique id or to use display_element with the unique id?
I am sorry for these basic questions. I really appreciate your help!

**Reply by vijaymarupudi** [reply]
All the modifications will be the in the plugin's code. To begin with, I would watch a tutorial on plugin creation, @jodeleeuw ran a tutorial on this here: <https://www.youtube.com/watch?v=LP7o0iAALik>

When you get the display_element, you'll have to do this:

```javascript
display_element.innerHTML = '<div id="distractor"></div><div id="main_task"></div>' 
display_element_distractor = document.querySelector('#distractor')
display_element_main_task = document.querySelector('#main_task')
```

You can then use these two display elements for the two different plugins you will be merging.

Don't worry, feel free to ask questions!

**Reply by jodeleeuw** [reply]
The plugin creation part of that video comes in the last hour or so. There's lots of helpful stuff before that, but if you are interested in just plugin creation I recorded the same tutorial as a standalone video: https://www.youtube.com/watch?v=XQcsFwAmbiw.

**Reply by avashipman** [reply]
Thank you both so much! I will definitely check out those videos!!

I also want to ask if it is possible to use the html-keyboard-response plugin with the survey-text plugin at the same time. I know I originally asked about the image-button-response, but I think the oddball task would work better with the keyboard response. I think this could be a bit tricky because the user is using the keyboard to answer the question.

**Reply by vijaymarupudi** [reply]
How are you envisioning they would the complete the oddball task and the keyboard response task at the same time?

**Reply by avashipman** [reply]
I am hoping to have the top of the screen be the survey-text part where the user answers the question and then below that would be the distractor. I was thinking of making the response to the target stimulus of the distractor be a number like 8 and then have the survey textbox not accept numbers so that only the distractor would receive the response.

**Reply by vijaymarupudi** [reply]
Ah I see. You'll need to listen to the [input](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) event, test whether the character that was inputted is a number, prevent the default action in that case, and then save the data and stop the oddball task.

I would focus on making the merged plugin barely functional before worrying about the semantics of this approach though!

**Reply by avashipman** [reply]
Ok sounds good! Thank you so much!!

**Comment by avashipman** [comment]
I created the plugin, but the distractor isn't showing up. The survey part is working though! I used your display_element method, but the two plugins (survey-text and image-keyboard-response) aren't sharing the screen. An error occurs when I run my experiment saying: Uncaught TypeError: Cannot set property 'innerHTML' of null. 

The error is only for the distractor. I tried taking out the height and width parameters to make it cleaner. I also tried making the ids the same and putting the display_element.innerHTML code in different orders in the plugin. Additionally, when I do this with the display_element_main_task.innerHTML it works, so I am not sure why the distractor isn't working.

This is the JavaScript code it is referring to:

      //display distractor stimulus
      var htmlDistractor = '<img src="'+trial.stimulus+'" id="distractor-stimulus" style="';
      if(trial.stimulus_height !== null){
        htmlDistractor += 'height:'+trial.stimulus_height+'px; '
        if(trial.stimulus_width == null && trial.maintain_aspect_ratio){
          htmlDistractor += 'width: auto; ';
        }
      }
      if(trial.stimulus_width !== null){
        htmlDistractor += 'width:'+trial.stimulus_width+'px; '
        if(trial.stimulus_height == null && trial.maintain_aspect_ratio){
          htmlDistractor += 'height: auto; ';
        }
      }
      htmlDistractor +='"></img>';
      display_element_distractor.innerHTML = htmlDistractor;   //<---- error here!

If you want, here is my plugin code: https://github.com/avashipman/Survey-Distractor.git

**Comment by vijaymarupudi** [comment]
Thank you for sending the code!

You've got a typo in the id portion of the HTML here

```js
display_element.innerHTML = '<div id="main_task-form"></div><div id="distactor-stimulus"></div>'
display_element_main_task = document.querySelector('#main_task-form')
display_element_distractor = document.querySelector('#distractor-stimulus')
```

You wrote distactor instead of distractor :)

**Reply by avashipman** [reply]
Thank you for seeing that mistake! So silly!! It worked!!!

Thank you again!

**Comment by avashipman** [comment]
Thank you for being so helpful! I really appreciate it.

I am still having trouble running the oddball task while the user is filling out the survey question. Any help you can give would be amazing! Thank you for all of your time already!

The oddball task uses the same structure as the demo experiment on jsPsych's tutorials. The images will change every 3 seconds and when the target image appears (at random), the user will press a key while the other images are NO_KEYS response. However, I don't want each image to be a new trial. I want the task to be running while the user is filling out the survey in 1 trial. I have tried to do this by setting the stimulus parameter as an array and change the image by going through the array. The only progress I have made is that when the images change they display down the page instead of replacing the old one. I am not sure how to go about it. I am also concerned that when I make the stimulus parameter an array, then it would be harder to specify the keyboard response for each image. This is the code I used to change the images in my plugin:

```javascript
// function to go to next stimulus
var next_stimulus = function() {

  // kill any remaining setTimeout handlers
  jsPsych.pluginAPI.clearAllTimeouts();

  // kill keyboard listeners
  if (typeof keyboardListener !== 'undefined') {
    jsPsych.pluginAPI.cancelKeyboardResponse(keyboardListener);
  }

  //next image
  function changeImage(){
    var img = document.getElementById('distractor-stimulus')
    var index = 0;
    index++;
    index = index % trial.stimulus.length; // This is for if this is the last image then goto first image
    img = trial.stimulus[index];
    };
  setInterval(changeImage, trial.response_next_image);

  if (response.key == trial.choices) {
    changeImage()
  }
}
```


This is the plugin code in its entirety: https://github.com/avashipman/Survey-Distractor.git

**Comment by vijaymarupudi** [comment]
Your `changeImage()` function isn't doing what you think it's doing. You are acquiring a reference to the `<img>` element and storing it in a variable, and then reassigning the variable to the value of your stimulus. These actions won't have an impact on that page.

You also define a new index that's zero every time the function is run. And you increment it right after, so `index` will also be equal to one. You probably want to move the `index = 0` line outside the function.

What you probably want is something like this:

```javascript
var index = 0;
var img = document.getElementById("distractor-stimulus");

function changeImage() {
  index++;
  index = index % trial.stimulus.length; // This is for if this is the last image then goto first image
  img.src = trial.stimulus[index];
}
```

**Reply by avashipman** [reply]
Thank you for your quick reply! I understand your advice which is helping me grasp and use JavaScript better so thank you for your help with these beginner mistakes. However, your suggestion hasn't seemed to work.

I tried putting the index and img variables outside of the function but it is saying that the src is null. I tried just making 
var img = trial.stimulus but that doesn't seem to work. And I also tried doing it like this in order to make the computer understand the array and directly display it, but it isn't setting the interval for the images on the screen:


```
//generate image order
    var image_order = [];
    for(var i=0; i<trial.stimulus.length; i++){
      image_order.push(i);
    }
    if(trial.randomize_image_order){
      image_order = jsPsych.randomization.shuffle(image_order);
    }

//display distractor stimulus
    var htmlDistractor = ' ';

    for (var i = 0; i < trial.stimulus.length; i++) {
      var images = trial.stimulus[ image_order[i] ];
      var image_index = image_order[ i ];
      htmlDistractor +='<div id="distractor-stimulus-'+image_index+''+images+'" class="distractor-stimulus-image">';
      var img = document.getElementById("distractor-stimulus");
      function changeImage() {
        image_index++;
        image_index = image_index % trial.stimulus.length; 
        img.src = trial.stimulus[image_index];
      }
      setInterval(changeImage, trial.response_next_image);
    
    htmlDistractor += '</div>';
    }

```
I would really appreciate any other suggestions/guidance. Thank you so much for your help and time!

**Reply by vijaymarupudi** [reply]
I think the problem here is that I don't know how you're using the plugin. Can you post code for that?

**Reply by vijaymarupudi** [reply]
I also think it's important to understand why `var img = trial.stimulus` won't do anything. You're declaring a new variable called `img` and assigning the value of `trial.stimulus` to it. You need to do something with `img` for an effect to be visible. I don't know what the data type of `trial.stimulus` is, so I'll wait for code on how it's being used!

**Reply by avashipman** [reply]
Thank you for explaining that!

Here is the code: https://github.com/avashipman/Survey-Distractor.git

**Reply by vijaymarupudi** [reply]
Ah this isn't what I asked for. I want to see how you're *using* it i.e. adding it to the timeline.

**Reply by avashipman** [reply]
I am so sorry! Here is how I am using it in the html file:

```javascript
var experiment = {
    timeline: [
    {
      type: "survey-distractor",
      questions: [{ prompt: "What did you do last weekend?" }],
      stimulus: jsPsych.timelineVariable('fileName'),
      choices: ['8'],
      button_label: "SUBMIT",
      preamble: "Please write one sentence.",
      response_next_image: 1000,
      data: jsPsych.timelineVariable('data'),
    		on_finish: function(data){
            data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
        },
    }
    ],
    timeline_variables: [
      {fileName: images[0], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[1], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[2], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[3], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[4], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[5], data: { correct_response: jsPsych.NO_KEYS }},
      {fileName: images[6], data: { correct_response: '8' }},
    ],
    randomize_order: true
  }
    timeline.push(experiment);

  jsPsych.init({
    timeline: timeline,
    on_finish: function () { jsPsych.data.displayData() },
  })
```

**Comment by vijaymarupudi** [comment, marked as answer]
So the problem here is that you want the images to change in the same trial, but by providing a single image per item in the timelineVariables array, each trial gets only one image.

For what you want, I believe you don't need timelineVariables for now. You need to pass in the array of images as `trial.stimulus`. I would remove the subtimeline and make it a normal trial too. Can you give it a shot then, with the code I posted here: https://github.com/jspsych/jsPsych/discussions/869#discussioncomment-41233

**Reply by avashipman** [reply]
I tried it and it is saying that the file is not found. I think it is because it is seeing the array as one picture source. So it is trying to find a picture that is named all the images.

**Reply by vijaymarupudi** [reply]
Can you post me code for how you're using it now? If you pass in the images as an array to trial.stimulus, and set the `img.src` to a member of the Array, that should work.

If possible, please post the entire experiment to the github repository, so that I can debug more throughly!

**Reply by avashipman** [reply]
Of course! Here is the link to both the javascript and the HTML: https://github.com/avashipman/Survey-Distractor.git

**Reply by vijaymarupudi** [reply]
* I removed preloading functionality from the plugin, since your use case is more complex, see manual preloading for how to preload. This was stopping the experiment from starting. jsPsych should do a better job of displaying the appropriate error message in this case.

* `changeImage()` was being called in a loop. It should be called outside of a loop (setInterval will run indefinitely). So I pulled it out.

* You were initially setting the img tag like this


```javascript
'<img src="' + trial.stimulus + '" id="distractor-stimulus-" style="';
```

Adding an array to a string causes javascript to do weird things (a historial artifact). I changed it to make it start with no `src` attribute.

Also notice how the id has a trailing hyphen. I don't think you meant to do this?

* Image tags do not need a closing tag

```javascript
    htmlDistractor += '"></img>';
```

* I'm not sure why you're looping here. You want to show all the images at the same time? Probably not. I deleted this part. 

```javascript
    for (var i = 0; i < trial.stimulus.length; i++) {
      var images = trial.stimulus[image_order[i]];
      var image_index = image_order[i];
      htmlDistractor +=
        '<div id="distractor-stimulus-' +
        image_index +
        "" +
        images +
        '" class="distractor-stimulus-image">';
      var img = document.getElementById("distractor-stimulus");
      var index = 0;

      htmlDistractor += "</div>";
    }
```

* There was code duplication for the setTimeout part, once using the jsPsych timer, and once using the browser default. I deleted the jsPsych timer part.

* You had a `<div>` with id `distractor-stimulus` and an `<img>` with the same id. The id attribute should be completely unique for the whole page. This is the **critical bug**, your `img` variable was referring to the `<div>`, not the image.


A few notes:

* The bug is probably not where you think it is. Providing a working copy helps me find the true bug :) The copies you provided were not working copies, I had to change file paths and find samples images to test the plugin. Tips for next time!

* I'm very impressed at the complexity of the code. This is a monumental task!

[Survey-Distractor.zip](https://github.com/jspsych/jsPsych/files/4974125/Survey-Distractor.zip)

**Reply by avashipman** [reply]
Thank you so so much! This is exactly what I wanted!! I also appreciate you explaining my mistakes and how you fixed them. This is extremely helpful as a beginner coder. Thank you so much again and you are the best!
